src/libs/elektra/hooks.c

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdbinternal.h>
+
+/**
+ * Uninitializes and frees all hooks in the passed KDB handle.
+ *
+ * @param kdb the KDB handle where the hooks should be freed
+ * @param errorKey the key which holds errors and warnings which were issued
+ */
+void freeHooks (KDB * kdb, Key * errorKey)
+{
+	if (kdb->hooks.gopts.plugin != NULL)
+	{
+		elektraPluginClose (kdb->hooks.gopts.plugin, errorKey);
+	}
+}
+
+static size_t getFunction (Plugin * plugin, const char * functionName, Key * errorKey)
+{
+	size_t result = elektraPluginGetFunction (plugin, functionName);
+
+	if (result == 0)
+	{
+		ELEKTRA_ADD_INSTALLATION_WARNINGF (errorKey, "Plugin '%s' does not implement function '%s'", plugin->name, functionName);
+	}
+
+	return result;
+}
+
+static int initHooksGopts (KDB * kdb, Plugin * plugin, Key * errorKey)
+{
+	if (!plugin)
+	{
+		return -1;
+	}
+
+	kdb->hooks.gopts.plugin = plugin;
+
+	if ((kdb->hooks.gopts.kdbHookGoptsGet = (kdbHookGoptsGetPtr) getFunction (plugin, "hook/gopts/get", errorKey)) == NULL)
+	{
+		return -1;
+	}
+
+	return 0;
+}
+
+static KeySet * getPluginConfigFromContract (const char * pluginName, const KeySet * contract)
```

Comment:
This function is complicated enough that it should probably have some documentation comment. Something simple like

```suggestion
/**
 * Extracts the config for a single plugin from the contract
 *
 * The config must be below system:/elektra/contract/mountglobal/<pluginName> and will be moved to below user:/
 *
 * @param pluginName The name of the plugin for which the config will be extracted
 * @param contract The contract from which the config is extracted
 */
static KeySet * getPluginConfigFromContract (const char * pluginName, const KeySet * contract)
```

Reply:
Yes, but on that topic - I'd like to create some tests for this function. Do you know of a way that I can use it within a test case without also exposing it through the binary? Currently it's static so no entry in the function table is created for it. 
I just don't want anyone outside of Elektra to call it.

I mean I could `#include` the .c file in the unit test ...

Reply:
There is some info about such tests in #4246 (not merged yet): 

https://github.com/ElektraInitiative/libelektra/blob/2e89bfdd38f6cefb0deb17d0e677385f92367b33/doc/decisions/header_file_structure.md?plain=1#L109-L119

You can try to follow this idea (probably needs new CMake setup), or you just do what we did until now and:
1. remove the `static`
2. Add the function name to `libelektraprivate_1.0` in the right `symbols.map` (in your case `src/libs/elektra/symbols.map`), to actually export the symbol
3. make sure the test links against the right library